pkg/resources/catalog.cattle.io/v1/clusterrepo/validator.go

```diff
@@ -0,0 +1,87 @@
+// Package clusterrepo is used for validating clusterrepo admission request.
+package clusterrepo
+
+import (
+	"fmt"
+
+	catalogv1 "github.com/rancher/rancher/pkg/apis/catalog.cattle.io/v1"
+	"github.com/rancher/webhook/pkg/admission"
+	v1 "github.com/rancher/webhook/pkg/generated/objects/catalog.cattle.io/v1"
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var gvr = schema.GroupVersionResource{
+	Group:    "catalog.cattle.io",
+	Version:  "v1",
+	Resource: "clusterrepos",
+}
+
+// NewValidator will create a newly allocated Validator.
+func NewValidator() *Validator {
+	return &Validator{}
+}
+
+// Validator conforms to the webhook.Handler interface and is used for validating request for clusterrepos.
+type Validator struct {
+	admitter admitter
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{
+		admissionregistrationv1.Update,
+		admissionregistrationv1.Create,
+	}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	return []admissionregistrationv1.ValidatingWebhook{*admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.NamespacedScope, v.Operations())}
+}
+
+// Admitters returns the admitter objects used to validate clusterrepos.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+type admitter struct {
+}
+
+// Admit is the entrypoint for the validator. Admit will return an error if it is unable to process the request.
+// If this function is called without NewValidator(..) calls will panic.
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	if request.Operation == admissionv1.Create || request.Operation == admissionv1.Update {
+		newClusterRepo, err := v1.ClusterRepoFromRequest(&request.AdmissionRequest)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get clusterRepo from request: %w", err)
+		}
+
+		err = a.validateFields(newClusterRepo)
+		if err != nil {
```

Comment:
Here, you'd need to return the main error from Admin based on what kind of error you got from `validateFields`.
You could do something like this: https://github.com/rancher/webhook/blob/e401287e9f7f36f1a3d718c68918b0a3b3227147/pkg/resources/management.cattle.io/v3/projectroletemplatebinding/validator.go#L105.
Or just always return `admission.ResponseBadRequest(err.Error())`, as in that example above.

Reply:
The bottom line is that if you get a normal validation error, the admission response is not nil, it just will indicate that the request is not allowed (the value Allowed on the response is false).

Reply:
Done